feat: Added the docs for datasources (#3704)
harness/developer-hub: docs/internal-developer-portal/features/checks-datasources.md

---

> title: Datasources

Reviewer: ```suggestion
title: Scorecard Data Sources
```

description: Adding Custom Checks and Data Sources for Scorecards 
sidebar_position: 4
helpdocs_topic_id:
helpdocs_category_id:
helpdocs_is_private: false
helpdocs_is_published: true
---

```mdx-code-block
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
```

> Harness IDP allows you to integrate various data sources and implement custom checks to ensure your software components adhere to best practices and compliance. In this docs, we'll walk through how to add custom checks and data sources for scorecards in Harness IDP.

Reviewer: Should link to the Scorecards main page for learning more about it.

### Overview

> Harness IDP allows you to integrate various data sources, such as GitHub, GitLab, Bitbucket, Azure DevOps, and many more, to collect specific data points for each software component. Once a data source is enabled, HDH will start storing data points from the source for each component.

Reviewer: ```suggestion
Harness IDP allows you to integrate various data sources, such as GitHub, GitLab, Bitbucket, Azure DevOps, and many more, to collect specific data points for each software component. Once a data source is enabled, you can use them to create checks to be used in scorecards.
```

### Add Custom Checks

> 1. Under the `Admin` go to `Scorecard`.

Reviewer: ```suggestion
1. Under the `Admin` go to `Scorecards`.
```

2. In the `Scorecards` go to the `Checks` tab and select **Create Custom Check**.
3. Now on the **Create Check** page add a name and description for your check.

> 4. Under **Rules** you can find the following Datasources to select from.

Reviewer: ```suggestion
4. Under **Rules** you can find the following Data Sources to select from.
```

:::info

> There's a tab called `Datasources` available in `Scorecards` page to check for supported datasources and the corresponding datapoints. 

Reviewer: ```suggestion
There's a tab called `Data Sources` available in `Scorecards` page to check for supported data sources and the corresponding data points. 
```

:::

```mdx-code-block
<Tabs>
<TabItem value="GitHub">
```

> The following **Datapoints** are available for GitHub. 

Reviewer: ```suggestion
The following **Data Points** are available for GitHub. 
```

1. **Branch Protection**
- *Objective:* Ensure that branch protection rules disallow force push and delete.
- *Calculation Method:* Fetch `backstage.io/source-location` annotation from the catalog YAML file to find repository details and verify the branch protection rules.
- *Prerequisites:* Github Connector with Admin access. Provide suitable `backstage.io/source-location` annotation if the catalog YAML file is present outside the source GitHub repository.

<docimage path={require('../../internal-developer-portal/features/static/branch-protection.png')}/>

**Example YAML**

> ```

Reviewer: ```suggestion
```yaml
```

**Example YAML**

```
kind: "Component"
apiVersion: "backstage.io/v1alpha1"
metadata:
  name: order-service
  annotations:
    backstage.io/source-location: 'url:https://github.com/kubernetes/kubernetes/tree/master'
    ...
spec:
    ...
```

3. **Mean Time to Merge Pull Request**
- *Objective*: Calculate the average time taken to merge the last 100 pull requests.
- *Calculation Method:* Retrieve repository details using `backstage.io/source-location` and calculate the average merge time.
- *Prerequisites:* Provide suitable `backstage.io/source-location` annotation if the catalog YAML file is present outside the source GitHub repository. Make sure to mention the branch name in the conditional input field.

<docimage path={require('../../internal-developer-portal/features/static/mean-time-github.png')}/>

**Example YAML**

```
kind: "Component"
apiVersion: "backstage.io/v1alpha1"
metadata:
  name: order-service
  annotations:
    backstage.io/source-location: 'url:https://github.com/kubernetes/kubernetes/tree/master'
    ...
spec:
    ...
```

```mdx-code-block
</TabItem>
<TabItem value="GitLab">
```

> The following **Datapoints** are available for GitLab. 

Reviewer: ```suggestion
The following **Data Points** are available for GitLab. 
```

**Example YAML**

```
kind: "Component"
apiVersion: "backstage.io/v1alpha1"
metadata:
  name: order-service
  annotations:
    backstage.io/source-location: 'url:https://gitlab.com/kubernetes/kubernetes/tree/master'
    ...
spec:
    ...
```

3. **Mean Time to Merge Pull Request**
- *Objective:* Calculate the average time taken to merge the last 100 pull requests.
- *Calculation Method:* Retrieve repository details using `backstage.io/source-location` and calculate the average merge time.
- *Prerequisites:* Provide suitable `backstage.io/source-location` annotation if the catalog YAML file is present outside the source GitLab repository. Make sure to mention the branch name in the conditional input field.

<docimage path={require('../../internal-developer-portal/features/static/m-t-gitlab.png')}/>

**Example YAML**

```
kind: "Component"
apiVersion: "backstage.io/v1alpha1"
metadata:
  name: order-service
  annotations:
    backstage.io/source-location: 'url:https://gitlab.com/kubernetes/kubernetes/tree/master'
    ...
spec:
    ...
```

```mdx-code-block
</TabItem>
<TabItem value="Bitbucket">
```

> The following **Datapoints** are available for Bitbucket

Reviewer: ```suggestion
The following **Data Points** are available for Bitbucket
```

**Example YAML**

```
kind: "Component"
apiVersion: "backstage.io/v1alpha1"
metadata:
  name: order-service
  annotations:
    backstage.io/source-location: 'url:https://bitbucket.org/kubernetes/kubernetes/tree/master'
    ...
spec:
    ...
```

2. **Mean Time to Merge Pull Request**
- *Objective:* Calculate the average time taken to merge the last 100 pull requests.
- *Calculation Method:* Retrieve repository details using `backstage.io/source-location` and calculate the average merge time.
- *Prerequisites:* Provide suitable `backstage.io/source-location` annotation if the catalog YAML file is present outside the source Bitbucket repository. Make sure to mention the branch name in the conditional input field.

<docimage path={require('../../internal-developer-portal/features/static/m-t-bitbucket.png')}/>

**Example YAML**

```
kind: "Component"
apiVersion: "backstage.io/v1alpha1"
metadata:
  name: order-service
  annotations:
    backstage.io/source-location: 'url:https://bitbucket.org/kubernetes/kubernetes/tree/master'
    ...
spec:
    ...
```

```mdx-code-block
</TabItem>
<TabItem value="Harness">
```
**Pre-Requisites**

> - For the functioning of Harness Data Source related checks the Harness NextGen CI/CD plugin should be configured with new annotations in catalog info YAML, `harness.io/pipelines` and `harness.io/services` as mentioned in the setup steps instruction of [Harness Nextgen CI/CD plugin](https://github.com/harness/backstage-plugins/tree/main/plugins/harness-ci-cd#harness-nextgen-cicd-plugin)

Reviewer: ```suggestion
- For the functioning of Harness Data Source related checks, the Harness CI/CD plugin should be configured with new annotations in catalog info YAML, `harness.io/pipelines` and `harness.io/services` as mentioned in the setup steps instruction of [Harness CI/CD plugin](https://github.com/harness/backstage-plugins/tree/main/plugins/harness-ci-cd#harness-nextgen-cicd-plugin)
```

:::info

In the Harness Data source, the first pipeline URL from `harness.io/pipelines` is considered for score computation and similarly, the first service URL from `harness.io/services` is considered.

:::

> The following **Datapoints** are available for Harness. 

Reviewer: ```suggestion
The following **Data Points** are available for Harness. 
```

<docimage path={require('../../internal-developer-portal/features/static/pager-catalog.png')}/>

**Example YAML**

```
kind: "Component"
apiVersion: "backstage.io/v1alpha1"
metadata:
  name: order-service
  annotations:
    pagerduty.com/service-id: PT5ED69
    ...
spec:
    ...
```

```mdx-code-block
</TabItem>
<TabItem value="Kubernetes">
```

### Prerequisites:

1. **Plugin Configuration**

- The Kubernetes plugin needs to be configured. Refer [here](https://developer.harness.io/docs/internal-developer-portal/plugins/available-plugins/kubernetes).

:::info

The plugin provides 2 ways to authenticate - `serviceAccount` and `Google authentication`. But currently, the scorecards supports `serviceAccount` type authentication only.

:::

2. **Entity Configuration**

- There are two ways to surface Kubernetes components as part of an entity using annotations - `backstage.io/kubernetes-id` and `backstage.io/kubernetes-label-selector`. But currently, scorecards support [backstage.io/kubernetes-label-selector](https://backstage.io/docs/features/kubernetes/configuration/#label-selector-query-annotation) annotation only. 

> The following **DataPoints** are avilable for Kubernetes Datasource.

Reviewer: ```suggestion
The following **Data Points** are available for Kubernetes Data source.
```

<docimage path={require('../../internal-developer-portal/features/static/replica-k8s.png')}/>

2. **Days passed since the application was last deployed**:

- *Objective:* Fetches the number of days since the most recent deployment was done.
- *Calculation Method:* The label selector configured in the catalog YAML is used to identify the `Kubernetes workload` and the `lastUpdateTime` is used from the conditions section. The cluster details configured in the Kubernetes plugin are used. If more than one cluster is configured, the workload search is done in each cluster and the oldest deployment time of all is taken into consideration.
- *Prerequisites:* The Kubernetes plugin needs to be configured and enabled in the admin section. Refer [here](https://developer.harness.io/docs/internal-developer-portal/plugins/available-plugins/kubernetes) for more details. 

<docimage path={require('../../internal-developer-portal/features/static/days-k8s.png')}/>

```mdx-code-block
</TabItem>
<TabItem value="Jira">
```

> The following **DataPoints** are avilable for Jira Datasource.

Reviewer: ```suggestion
The following **Data Points** are avilable for Jira Datasource.
```
